File: 0463-island-perimeter/0463-island-perimeter.cpp
class Solution {
public:
    int dx[4] = {-1, 0, 1, 0};
    int dy[4] = {0, 1, 0, -1};
    
    bool isInBoundary(int &x, int &y, int &height, int &width) {
        if (0<= x && x < height && 0 <= y && y < width) {
            return true;
        }
        return false;
    }
    
    int numberOfBoundary(vector<vector<int>> &grid, int &x, int &y) {
        int result = 0;
        int height = grid.size();
        int width = grid[0].size();
        
        for (int dir = 0; dir < 4; dir++) {
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            
            if (!isInBoundary(nx, ny, height, width)) {
                result++;
                continue;
            }
            if (grid[nx][ny] == 0) {
                result++;
                continue;
            }
        }
        
        return result;
    }
    
    int islandPerimeter(vector<vector<int>>& grid) {
        pair<int, int> start;
        // find the first point to start BFS
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 1) {
                    start = {i, j};
                    break;
                }
            }
        }
        int height = grid.size();
        int width = grid[0].size();
        vector<vector<bool>> isVisited(height, vector<bool>(width, false));
        
        int result = 0;
        queue<pair<int, int>> q;
        q.push(start);
        isVisited[q.front().first][q.front().second] = true;
        int dx[4] = {-1, 0, 1, 0};
        int dy[4] = {0, 1, 0, -1};
        
        while(!q.empty()) {
            int cx = q.front().first;
            int cy = q.front().second;
            q.pop();
            
            result += numberOfBoundary(grid, cx, cy);
            
            for (int dir = 0; dir < 4; dir++) {
                int nx = cx + dx[dir];
                int ny = cy + dy[dir];

                if (!isInBoundary(nx, ny, height, width) || grid[nx][ny] == 0 || isVisited[nx][ny]) {
                    continue;
                }
                isVisited[nx][ny] = true;
                q.push({nx, ny});
            }
        }
        
        return result;
    }
};

File: 0463-island-perimeter/0463-island-perimeter.kt
class Solution {
    
    val dx: Array<Int> = arrayOf(-1, 0, 1, 0)
    val dy: Array<Int> = arrayOf(0, 1, 0, -1)
    
    fun isInBoundary(x: Int, y: Int, width: Int, height: Int): Boolean = 
        if (0 <= x && x < width && 0 <= y && y < height) true
        else false
    
    fun numberOfBoundary(grid: Array<IntArray>, x: Int, y: Int): Int {
        var result = 0
        val height = grid.size
        val width = grid[0].size
        
        for (dir in 0 until 4) {
            val nx = x + dx[dir]
            val ny = y + dy[dir]
            
            if (!isInBoundary(nx, ny, height, width) || grid[nx][ny] == 0) {
                result++
            }
        }
        
        return result
    }

    fun islandPerimeter(grid: Array<IntArray>): Int {
        var q: Queue<Pair<Int, Int>> = LinkedList()
        // find the first point to start BFS
        for (i in 0 until grid.size) {
            var temp: Boolean = false
            for (j in 0 until grid[0].size) {
                if (grid[i][j] == 1) {
                    q.add(Pair(i, j))
                    temp = true
                    break
                }
            }
            if (temp) break
        }
        
        val height = grid.size
        val width = grid[0].size
        val isVisited = Array(height) { BooleanArray(width) }
        
        var result: Int = 0
        isVisited[q.peek().first][q.peek().second] = true
        
        while(!q.isEmpty()) {
            val cx = q.peek().first
            val cy = q.peek().second
            q.remove()
            
            result += numberOfBoundary(grid, cx, cy)
            
            for (dir in 0 until 4) {
                val nx = cx + dx[dir]
                val ny = cy + dy[dir]
                
                if (!isInBoundary(nx, ny, height, width)) {
                    continue
                }
                if (grid[nx][ny] == 0 || isVisited[nx][ny]) {
                    continue
                }
                isVisited[nx][ny] = true
                q.add(Pair(nx, ny))
            }
        }
        
        return result
    }
}